cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -388,31 +389,44 @@ ignoringSeverity tr = Tracer $ \(WithSeverity _ ev) -> traceWith tr ev
 
 traceChainMetrics
   :: forall blk. HasHeader (Header blk)
-  => Trace IO Text -> Tracer IO (ChainDB.TraceEvent blk)
-traceChainMetrics tr = Tracer $ \ev ->
-  fromMaybe (pure ()) $
-    doTrace <$> chainTipInformation ev
- where
-   chainTipInformation :: ChainDB.TraceEvent blk -> Maybe ChainInformation
-   chainTipInformation = \case
-     ChainDB.TraceAddBlockEvent ev -> case ev of
-       ChainDB.SwitchedToAFork _warnings newTipInfo _ newChain ->
-         Just $ chainInformation newTipInfo newChain
-       ChainDB.AddedToCurrentChain _warnings newTipInfo _ newChain ->
-         Just $ chainInformation newTipInfo newChain
-       _ -> Nothing
-     _ -> Nothing
-
-   doTrace :: ChainInformation -> IO ()
-   doTrace ChainInformation { slots, blocks, density, epoch, slotInEpoch } = do
-     -- TODO this is executed each time the chain changes. How cheap is it?
-     meta <- mkLOMeta Critical Public
-
-     traceD tr meta "density"     (fromRational density)
-     traceI tr meta "slotNum"     slots
-     traceI tr meta "blockNum"    blocks
-     traceI tr meta "slotInEpoch" slotInEpoch
-     traceI tr meta "epoch"       (unEpochNo epoch)
+  => Trace IO Text
+  -> IO (Tracer IO (ChainDB.TraceEvent blk))
+traceChainMetrics tr = do
+  tSwitchingBlocksGained <- STM.newTVarIO 0
+  tSwitchingBlocksLost <- STM.newTVarIO 0
+  return . Tracer $ \ev ->
+    fromMaybe (pure ()) $
+      doTrace tSwitchingBlocksGained tSwitchingBlocksLost <$> chainTipInformation ev
+  where
+    chainTipInformation :: ChainDB.TraceEvent blk -> Maybe ChainInformation
+    chainTipInformation = \case
+      ChainDB.TraceAddBlockEvent ev -> case ev of
+        ChainDB.SwitchedToAFork _warnings newTipInfo _ chain ->
+          Just $ chainInformation newTipInfo chain 0 0
```

Comment:
I need to insert some logic here to count number of blocks lost and gained due to switching.  Also need a way to tell which blocks are mine.

Reply:
```
-- | The new block fits onto some fork and we have switched to that fork
-- (second fragment), as it is preferable to our (previous) current chain
-- (first fragment).
SwitchedToAFork
  [LedgerEvent blk]
  (NewTipInfo blk)
  (AnchoredFragment (Header blk))
  (AnchoredFragment (Header blk))
  ```
  
  So you want this pattern:
  
  ```
  ChainDB.SwitchedToAFork _warnings newTipInfo headersLost headersGained
  ```

Reply:
Off the top of my head, I don't think the `HasHeader (Header blk)` constraint here is enough to determine directly if the header was one you forged or not. Maybe indirectly, eg if that list of points you forged is already being maintained nearby.

Reply:
@nc6 informs me that I can get it from here:

https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/BlockChain.hs#L412

I suppose I need extra constraints to access these?

Reply:
I feel like I need something similar to this type family:

https://github.com/input-output-hk/ouroboros-network/blob/b3730282ea065d0f902cac1dcb3237e01fecd1cf/ouroboros-consensus/src/Ouroboros/Consensus/Protocol/Signed.hs#L12

It's not clear to me there is enough to do this without changes to `ouroboros-network`.

Reply:
Also, where can I get my own keys that I can compare to the ones in the block?